File: app/src/main/java/com/carloid/nitromemorybooster/view/fragment/MainFragment.java
package com.carloid.nitromemorybooster.view.fragment;

import android.animation.ObjectAnimator;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateDecelerateInterpolator;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;

import com.carloid.nitromemorybooster.R;
import com.carloid.nitromemorybooster.databinding.FragmentMainBinding;
import com.carloid.nitromemorybooster.util.U;

import java.util.Locale;
import java.util.Objects;

import static android.Manifest.permission.READ_EXTERNAL_STORAGE;
import static android.Manifest.permission.WRITE_EXTERNAL_STORAGE;

public class MainFragment extends Fragment {

    private static final int PERMISSION_REQUEST = 10001;
    private static final String[] PERMISSION_LIST = { READ_EXTERNAL_STORAGE, WRITE_EXTERNAL_STORAGE };

    private FragmentMainBinding binding;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        requestPermissions(PERMISSION_LIST, PERMISSION_REQUEST);

        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_main, container, false);
        components();

        U.listDirContents(Environment.getExternalStorageDirectory());

        return binding.getRoot();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == PERMISSION_REQUEST) {
            for (int grantIterator : grantResults) {
                if (grantIterator != PackageManager.PERMISSION_GRANTED) {
                    Objects.requireNonNull(getActivity()).finish();
                }
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    private void components() {
        final float usable = getUsableMemory();
        final float free = getFreeMemory();
        final int percentage = (int)(100.0f * (usable - free) / usable);

        binding.txtUsable.append(String.format(Locale.getDefault(), "%.1f GB", usable));
        binding.txtFree.append(String.format(Locale.getDefault(), "%.1f GB", free));
        binding.txtOSUsage.append(String.format(Locale.getDefault(), "%.1f GB", getSystemMemoryUsage()));

        final String percentStr = String.format(Locale.getDefault(), "%d", percentage) + "%";
        binding.percentage.setText(percentStr);

        ObjectAnimator progressAnim = ObjectAnimator.ofInt(binding.progressBar, "progress", 0, percentage);
        progressAnim.setDuration(1000);
        progressAnim.setInterpolator(new AccelerateDecelerateInterpolator());
        progressAnim.start();

        binding.cleanBtn.setOnClickListener(v -> {
            Log.d("TEST", "Click");
        });
    }

    private float getUsableMemory() {
        return U.getPathTotalSpace(Environment.getExternalStorageDirectory().getPath());
    }

    private float getFreeMemory() {
        return U.getPathFreeSpace(Environment.getExternalStorageDirectory().getPath());
    }

    private float getSystemMemoryUsage() {
        float product = U.getPathTotalSpace("/product");
        float system  = U.getPathTotalSpace("/system");

        return product + system;
    }

    private float getImageFilesUsage() {
        return U.measureFileTypeUsage(".png", ".bmp", ".jpg", ".jpeg", ".gif", ".webp");
    }

    private float getAudioFilesUsage() {
        return U.measureFileTypeUsage(".wav", ".mp3", ".ogg", ".m4a", ".aac", ".flac", ".mid", ".midi");
    }

    private float getDocumentFilesUsage() {
        return U.measureFileTypeUsage(".doc", ".docx", ".pdf", ".ppt", ".pptx");
    }

    private float getVideoFilesUsage() {
        return U.measureFileTypeUsage(".3gp", ".mp4", ".mkv", ".webm");
    }
}

File: app/src/main/java/com/carloid/nitromemorybooster/view/fragment/MainFragment.kt
package com.carloid.nitromemorybooster.view.fragment

import android.Manifest.permission
import android.animation.ObjectAnimator
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AccelerateDecelerateInterpolator
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import com.carloid.nitromemorybooster.R
import com.carloid.nitromemorybooster.databinding.FragmentMainBinding
import com.carloid.nitromemorybooster.util.KotlinUtil.Companion.listDirObjects
import com.carloid.nitromemorybooster.util.U
import java.io.File
import java.util.*

class MainFragment : Fragment() {

    private var binding: FragmentMainBinding? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        requestPermissions(PERMISSION_LIST, PERMISSION_REQUEST)

        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_main, container, false)
        components()

        return binding!!.root
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == PERMISSION_REQUEST) {
            for (grantIterator in grantResults) {
                if (grantIterator != PackageManager.PERMISSION_GRANTED) {
                    Objects.requireNonNull(activity)!!.finish()
                }
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    private fun components() {
        val usable = usableMemory
        val free = freeMemory
        val percentage = (100.0f * (usable - free) / usable).toInt()

        binding!!.txtUsable.append(String.format(Locale.getDefault(), "%.1f GB", usable))
        binding!!.txtFree.append(String.format(Locale.getDefault(), "%.1f GB", free))
        binding!!.txtOSUsage.append(String.format(Locale.getDefault(), "%.1f GB", systemMemoryUsage))

        val percentStr = String.format(Locale.getDefault(), "%d", percentage) + "%"
        binding!!.percentage.text = percentStr

        val progressAnim = ObjectAnimator.ofInt(binding!!.progressBar, "progress", 0, percentage)
        progressAnim.duration = 1000
        progressAnim.interpolator = AccelerateDecelerateInterpolator()
        progressAnim.start()

        binding!!.cleanBtn.setOnClickListener {
            Log.d("TEST", "Click")
        }

        Handler(Looper.getMainLooper()).postDelayed({
            var externalDir = File(EXTERNAL_STORAGE_DIR)
            while (!externalDir.isDirectory) {
                externalDir = File(EXTERNAL_STORAGE_DIR)
            }

            // Image
            val imageList = Vector<String>()
            val imageFormats = listOf(".png", ".bmp", ".gif", ".webp")

            listDirObjects(externalDir, imageList, 0.0f, imageFormats) { size ->
                binding!!.txtUsedImages.text = String.format("%.1f GB", U.bytesToGigaBytes(size))
            }

            // Documents
            val documentList = Vector<String>()
            val docFormats = listOf(".pdf", ".doc", ".ppt", ".docx", ".pptx")

            listDirObjects(externalDir, documentList, 0.0f, docFormats) { size ->
                binding!!.txtUsedDocuments.text = String.format("%.1f GB", U.bytesToGigaBytes(size))
            }

            // Videos
            val videoList = Vector<String>()
            val vidFormats = listOf(".mp4", ".3gp", ".webm")

            listDirObjects(externalDir, videoList, 0.0f, vidFormats) { size ->
                binding!!.txtUsedVideos.text = String.format("%.1f GB", U.bytesToGigaBytes(size))
            }

            // Audios
            val audioList = Vector<String>()
            val audFormats = listOf(".mp3", ".ogg", ".wav")

            listDirObjects(externalDir, audioList, 0.0f, audFormats) { size ->
                binding!!.txtUsedAudios.text = String.format("%.1f GB", U.bytesToGigaBytes(size))
            }
        }, 500)
    }

    private val usableMemory: Float
        get() = U.getPathTotalSpace(EXTERNAL_STORAGE_DIR)

    private val freeMemory: Float
        get() = U.getPathFreeSpace(EXTERNAL_STORAGE_DIR)

    private val systemMemoryUsage: Float
        get() {
            val product = U.getPathTotalSpace("/product")
            val system = U.getPathTotalSpace("/system")
            return product + system
        }

    companion object {
        private const val PERMISSION_REQUEST = 10001
        private val PERMISSION_LIST = arrayOf(permission.READ_EXTERNAL_STORAGE, permission.WRITE_EXTERNAL_STORAGE)
        private const val EXTERNAL_STORAGE_DIR = "/storage/emulated/0"
    }
}